### DISCUSSION TITLE
Super Simple Bronze Mode Conformance

### DISCUSSION THREAD

**Discussion by Myndex** [AccessibleReadingTechnologies/APRG#5] [Ideas]
The very early drafts involving SAPC and later APCA involved the use of a font table that some found daunting. In an effort to bring perceptual uniformity to a wider group of users, we've worked to find the bare essentials, the **absolute minimums that would still be useful** as a **most-basic** and easiest-to-implement **guideline**.

-----
### The Irreducibly Short Bronze Simple Level
_This is intended to be visible on or near a tool's UI as a quick reference._

**Lc 15:** Thick Non-Text
**Lc 30:** Sub-fluent Text
**Lc 45:** Large Text  > 36px
**Lc 60:** Fluent Text  > 24px
**Lc 75:** Fluent / Body Text  > 18px
**Lc 90:** Preferred Body Text  > 16px
_(Sizes per a 0.52 x-height ratio)_

### The Slightly Embellished Bronze Simple Level
_(For tooltips or when there is more space available)_

**Lc 15:** Large / Thick Non-Text (no less than 5px thick)
**Lc 30:** Sub-fluent Text min / semantic non-text min
**Lc 45:** Large Text  > 36px normal, 24px bold
**Lc 60:** Fluent Text  > 24px normal, 16px bold
**Lc 75:** Fluent / Body Text  > 18px normal
**Lc 90:** Preferred Body Text  > 16px

_Font sizes shown are for the reference fonts Arial or Helvetica._
_Other fonts should be compared and adjusted for equivalence._

------
## _BRONZE SIMPLE LEVEL IN DETAIL_
Edited Aug 20 2023 to indicate more font sizes and weights at each threshold.

Please comment if you feel this over complicates.

Bronze Simple Mode
- Uses a few basic threshold levels only, similar to WCAG_2.
- Makes polarity _optional_, provided that the algorithm is set to **_non-polar mode_**.
- Focuses on **_primary content text_**, ignoring things like "disabled controls" in the same way WCAG2 does.

**Bronze Use Case Conformance Values for Testing**
_**Lc** values listed using an approved APCA compliant algorithm_

### APCA-RC Bronze Simple Level Conformance

- **Body Text**
    - **> Lc ±75**
        - **> 18px** Normal 400 (24px 300, 16px 500)
    - _Preferred_: **Lc ±90**
        - **> 16px** Normal 400 (21px 300, 14px 600)
- **Small Fluent Text**¹
    - **> Lc ±75**
        - **> 16px** Normal 400 (24px 300, 15px 600, 14px Bold 700)
    - _Preferred_: **Lc ±90**
        - **> 14px** Normal 400²
- **Medium Fluent Text, small/thin Non-Text**
    - **> Lc ±60**
        - **> 24px** Normal 400 (36px 300, 21px 500, 18px 600, 16px Bold 700)
- **Large Fluent Text, semantic non-text**
    - **> Lc ±45**
        - **> 36px** Normal 400 (60px 300, 32px 500, 28px 600, 24px Bold 700)

¹ _Small Fluent Text means content that is not otherwise body text_
² _Recommended minimum for Lc90, not mandated_

- **Informative (Non-Normative) Recommendations**
    - **Subtract Lc 15 for Sub-fluent**, secondary-content text (footer, captions, small logos)
    - **Subtract Lc 30 for incidental** spot, non-content text (inactive, copyright, placeholder, large logos)
        - no text conveying meaning may be less than Lc 30.
        - Lc 15 is the invisibility point, especially for small or thin elements.
    - **No font-smoothing** (set to auto only) for fonts smaller than 24px with weight less than 600
    - **Maximum Contrast:** Lc -90 (for dark mode only) for bold or large fonts.
    - **Font size and weights as shown based on the reference** fonts Arial or Helvetica
        - Other reference fonts will be added to the reference list
        - A list of fonts that depart from the reference fonts is also being developed, indicating the needed offsets for best use.

------
### The APCA tool will soon output a line weight value in addition to Lc value for non text and dataviz.

|Semantic Non-Text<br>DataViz A|DataViz B| Simple Non-Semantic<br>Non-Text & DataViz C|
|:----:|:----:|:----:|
|Icons, pictograms<br>Complex Data Display|Chart Lines, Legends,<br>Pie Pieces, Bars, Maps|Meta States, Form Outlines<br> Chart Grids, Lead lines|
| <img width="100" alt="NonText Semantic 2023-04-11 at 6 26 55 PM" src="https://github.com/A11yReadTech/ARC/assets/42009457/99bb3f46-1d41-421d-90df-e388a5b22b6c"> |Currently In<br> Development| <img width="100" alt="Non Text Non Sem 2023-04-11 at 8 55 53 PM" src="https://github.com/A11yReadTech/ARC/assets/42009457/a65fdfda-83fa-467c-8aca-1c3882afdd06"> |

------

### _Understanding Bronze Simple Mode:_
- Covers primary content text.
- Use-cases for "body text" and "other content text" (definitions below)
- Spot-read text: disabled, ancillary, decorative text are _excluded_, but with recommended levels
- Logos and _incidental_ text in images are _excluded_, but with recommended levels
- There is **no font lookup table**
    - only the set of general conformance _levels_
    - A lookup table or tool is optional
- There is no _specified_ minimum font size for Lc&nbsp;90, but with recommended minimums
    - other thresholds do have basic font size minimums
    - There is no requirement to match a font to a reference font weight or size, but strongly recommended to do so.
- user zoom must not be restricted by content author, and must not create a blocking condition
- Polarity is _optional_ when:
    - The APCA background input is always the lightest color
        - so that only positive Lc values are output.
    - Or, the APCA algorithm is set to **_non-polar mode_**
        - Note that APCA non-polar mode reports the **_lowest_** contrast for any given polarity or spatial frequency.
    - Or when using the DPS Contrast algorithm (Delta Phi Star, which is inherently non-polar)

------
### ARC Use Case Definitions for Fluently Readable Text

**FLUENT**     
- Fluent Readability, **Block/Body Text**
    - **Defined as:** a _block_ or _column_ of three¹ or more _full_ ² continuous lines of content text in a block or column, that uses a readable font, and at a font size less than 36px.
- Fluent Readability, **Primary content** that is not body text
    - **Defined as:** up to two and a half lines of continuous text.
        - Can includes headlines, captions, and images of text, if they contribute to content
        - primary navigation and primary menus
        - asides, tool-tips, spoilers, "continued"
        - user entered text in forms, fields, etc
        - does not include text the user controls or adjusts in size or color
- Large Fluent **header/title content**
    - **Defined as:** fluent subcategory of "large content" such as big, bold headlines, and generally referring to text larger than 36px.

**_Note:_**    
¹ _Partial lines do not count toward the body-text block definition, so 2 ½ lines is not necessarily body text, unless it is a 2½ line paragraph as part of a larger block or column. Three full lines generally would be the bright line for body text._     
² _A full line is defined as a line of text at the average line width for the given text block, usuallhy no less than 35 characters, and preferably less than 60 characters per line average._

**Comment by lloydk** [comment]
Is there an a more detailed explanation of non-polar mode?

> - There is no specified minimum font size for Lc 90
    - thresholds under Lc 75 do have basic font minimums

Is there a minimum font size for Lc < 90 and >= 75?

> - Defined as: a block or column of more than two full continuous lines of content text that uses a readable font with an x-height of ~19px or less.

I think using the x-height of a font might be a barrier to simplicity and ease of use.

**Reply by Myndex** [reply]
Hi @lloydk Thank you for your comments:

> _"...Is there an a more detailed explanation of non-polar mode?..."_

It's pre-release, but basically, it is essentially the result that you get when you always calculate the text as darker than the background (in other words, the darkest color always goes into the text input regardless of how that color is being used).

It may be such that it checks both polarities and then always reports the lowest contrast. The idea behind non-polar mode is to simplify.


> _"...Is there a minimum font size for Lc < 90 and >= 75?..."_

We haven't quite settled, but are leaning toward 18px. We almost could say that 16px is possible, except when you look at a font such as RALEWAY, its 400 weight is equivalent of 300 weight vs the reference font.

And since _**Bronze Simple**_ is intended to be as simple as WCAG2, we are struggling with not using a reference font for a comparison, yet still providing meaningful results.

And actually I'd like to hear commentary regarding using minimum font sizes for Lc 75+: 

- **Choice 1:**
    - 14px bold, 18px normal, 24px 300. 

- **Choice 2:**
    - 14px bold, 16px normal, 21px 300. 

- **Choice 3:**
    - 12px bold, 14px normal, 18px 300. 

Choice one is what I would prefer.     
Choice three (or worse) is what a lot of people are doing (including some accessibility advocates).    
**Choice 2 splits the difference. **


> _"...I think using the x-height of a font might be a barrier to simplicity and ease of use..."_

Thank you for pointing that out, that was copy-pasta from the master use-case description, edited now for simplicity.

**Comment by lloydk** [comment]
If Silver, Gold, etc. keep the concept of polarity then I think overall the spec will be more confusing than if polarity was either in all conformance levels or ditched entirely. I'd prefer to see it kept in all levels without mentioning the word polarity which is a technical term that some people might not understand. 

I think that the negative values returned for WoB could be considered confusing so returning a contrast value >= 0 for BoW and WoB might help to simplify the concepts surrounding the APCA algorithm. Personally, I've never had a need to know what the polarity of the contrast value is and the fact that the value might be negative is mostly an inconvenience.

**Reply by Myndex** [reply]
> _,,,I'd prefer to see it kept in all levels without mentioning the word polarity,,,_

Hi @lloydk Thank you for commenting, non-polar mode it's literally brand new and experimental, it was an idea we started playing with very recently, looking for ways to find the _absolute simplest_ reduction to practice. 

That said, In order to get the benefits of **design flexibility** and **best readability** for silver & gold levels, polarity definitely needs to be a part of that.

Instead of the term **_polarity_** is there another term that you would suggest? 
- Mode as in light mode dark mode is a term that we've been using a lot,
- and then media queres uses the term color scheme. 

If you have thoughts here we'd love to hear them.



> _...I think that the negative values returned for WoB could be considered confusing..._

The substantial majority of users seem to prefer showing polarity as positive or negative, but I have noticed some edge cases with people who have never worked with APCA before being _initially_ confused, which is why I did that poll on Twitter. But this doesn't happen very often. And we've improved and simplified the documentation.

### quick question for _you_
And that brings up a point I'm very curious about, what if any of the documentation have you happened to read? And which tool or tools are you using?


> _...Personally, I've never had a need to know what the polarity..._

I don't expect designers Will necessarily need to know the polarity. Coming from a design point of view, particularly if you're using a design tool like Leonardo, as long as you know **which color is designated text** and which color is designated background.

As I indicated on that thread in Twitter, the reason for using a negative number is partly because it's actually natural to how the math _naturally_ functions. But in the written specification I also point out that what's important is that the text color and the background color are **_properly identified_**.

One way to figure out which color is which, if the number is negative then you know that the background is to be darker than the foreground, and vice versa.


### Testers

But while a designer using a design tool may not necessarily need to know what the polarity actually is, somebody who is doing **testing** of a site to see if it is conformant is most likely going to want to see polarity as a function of testing the site.

**Comment by lloydk** [comment]
I think by trying to simplify bronze level by introducing non-polar mode you're going to end up introducing more complexity in the the entire accessibility ecosystem (specs, documentation, tooling, etc.).

Not a big fan of the light mode/ dark mode terms for polarity. Those terms are commonly used to describe css color schemes. Just google `css light mode dark mode` and you'll see an endless number of articles and videos about creating light mode or dark mode color schemes/themes. So I think talking about polarity in terms of light/dark mode will be confusing. 

> And that brings up a point I'm very curious about, what if any of the documentation have you happened to read? And which tool or tools are you using?

I've read all of your documentation and articles many times over. I've tested hundreds of color combinations using your APCA contrast tool. Developed a bunch of small sites trying to conform to the  suggested APCA contrast ranges and font sizes. Implemented the APCA contrast function in GLSL and developed some APCA/WCAG comparison tools which may or may not get released publicly.

**Reply by Myndex** [reply]
> _I think by trying to simplify bronze level by introducing non-polar mode you're going to end up introducing more complexity in the the entire accessibility ecosystem (specs, documentation, tooling, etc.)._

Hi @lloydk 

Thank you for that feedback, this is the kind of feedback I wish we had more of… As I said we just brought up the idea of a non-polar mode, and we haven't even put it into any tools or code yet for testing and experimentation.

**_The reason:_** we are faced with a number of challenges, those challenges include integrating methods into existing testing tooling, and finding entry points for sites that are currently using WCAG&nbsp;2 to make it easy to shift over.

Also there are there legitimate cases for contrast between two elements that are not text or at high spatial frequencies. In these cases, a non-polar approach may provide certain benefits, and that also is part of why we were exploring some of these options.


> _I've read all of your documentation and articles many times over. I've tested hundreds of color combinations using your APCA contrast tool. Developed a bunch of small sites trying to conform to the suggested APCA contrast ranges and font sizes._ 

Excellent, thank you for letting me know.

Regarding your comments here or on Twitter for polarity indication, do you find anything that you do is being blocked or difficult because of the current polarity indicator? (i.e. minus for negative polarity), or is it just that you feel it's purpose is not clear? 

Regarding the term "polarity", you mentioned concerns about this as a term, is there another term that you think would be more appropriate?

Can you share any links to any of the sites that you put up?

**Comment by lloydk** [comment]
> Regarding your comments here or on Twitter for polarity indication, do you find anything that you do is being blocked or difficult because of the current polarity indicator? (i.e. minus for negative polarity), or is it just that you feel it's purpose is not clear?

I would say that it's clear (at least to me) what information the polarity indicator provides but as far as I can tell there's currently little if any use for it. The guidelines produced so far (the contrast ranges for various use cases) don't mention it so as I said on Twitter I'm not sure what I'm supposed to do with the polarity indicator. I say that as a user of your online contrast calculator, a developer of tools and a tester. In all of those scenarios I already know which colors are foreground and background and don't really care if the foreground is lighter or darker than the background. 

You've alluded to upcoming developments that make the polarity indicator more relevant so maybe when that work comes I'll have to take the indicator into account when checking against contrast ranges for various use cases. 

> Regarding the term "polarity", you mentioned concerns about this as a term, is there another term that you think would be more appropriate?

My only concern with the term "polarity" is that in a spec with diverse audience it might be to technical but I don't have any other suggestions. The main thought I had was the opportunity to simplify things by removing the polarity indicator entirely but it sounds like that won't be possible due to the upcoming changes.

**Reply by Myndex** [reply]
> _...guidelines produced so far ... don't mention it..._

Ah, I see, this is why I asked which documentation you've read thus far, as this sounds like a documentation issue... Documentation for this project is still a work in progress like several other things, so that's something that needs to be discussed I think in a little bit more detail.

Thank you

**Comment by Myndex** [comment]
## Update

Just updated the levels as shown below, above however, the descriptions were made more terse.

**_Note this is not "WCAG 2 backwards mode"._**

### APCA-RC Bronze Simple Level Conformance

- **Body Text**
    - Minimum Contrast: Lc ±75
        - Reference Font > ±16px 500 weight
        - Reference Font > ±18px 400 weight
        - Reference Font > ±24px 300 weight
    - Preferred Contrast: Lc ±90
        - Reference Font > ±14px 600 weight
        - Reference Font > ±16px 400 weight
        - Reference Font > ±21px 300 weight
- **Small Non-Body Text**
    - Minimum Contrast: Lc ±75
        - Reference Font > ±16px 400 weight
    - Preferred Contrast: Lc ±90
        - Reference Font > ±14px 400 weight
- **Medium Content Text, any Non-Text**
    - Minimum Contrast: Lc ±60
        - Reference Font > ±24px Normal or ±16px Bold
- **Large Content Text, **
    - Minimum Contrast: Lc ±45
        - Fonts larger than ±36px Normal or ±24px Bold


- **Informative (Non-Normative) Recommendations**
    - Subtract Lc 15 from above for Sub-fluent, secondary-content text (footer, captions, small logos)
    - Subtract Lc 30 for incidental spot, non-content text (inactive, copyright, placeholder, large logos)
        - no text conveying meaning may be less than Lc 30.
        - no semantic non-text (icons, pictograms) may be less than Lc 30.
        - Lc 15 is the invisibility point, especially for small or thin elements.
    - Maximum Contrast: Lc -90 (for dark mode only) for bold or large fonts.
    - Font size and weights as shown based on the reference fonts Arial or Helvetica
        - Other references font will be added to the reference list
        - A list of fonts that depart from the reference fonts is also being developed, indicating the needed offsets for best use.

**Comment by lloydk** [comment]
Are the font sizes really `>` or should they be `>=` ?

What does the the `±` symbol in front of the font sizes mean?

**Reply by Myndex** [reply]
Hi @lloydk 

Yes, >= in theory per the reference font—I am still taking feedback regarding reference fonts and font matching for Bronze level.

The ± means the Lc value for either positive or negative mode. There are cases where there will be different threshold for positive or negative polarity for certain features.

And incidentally, as I think you had asked regarding importance of marking polarity: we discovered it applies directly to **_maximum contrast_**.

**Lc -90 is maximum contrast**, and because it is negative, it applies only to dark mode polarities.

While there is no maximum contrast for light mode, there is a _maximum suggested brightness for backgrounds_ in light mode which will be discussed in the not too distant future.

**Comment by lloydk** [comment]
> And incidentally, as I think you had asked regarding important of marking polarity: we discovered it applies directly to **_maximum contrast_**.
> 
> **Lc -90 is maximum contrast**, and because it is negative, it applies only to dark mode polarities.
> 

Does the dark mode maximum contrast apply to all use cases or only to selected use cases like body text? There's a fair number of colors that exceed Lc -90 with a white foreground color. I can see darker buttons with a white foreground color failing because they're > Lc -90 being an issue for some people.

I just checked a few sites that I worked on and I have some buttons that are > Lc -90. I also have a lot of dark mode footers with 14px text that are > Lc -90. Lowering the contrast of one of the footers from Lc -101 to Lc 90 does in fact make it more readable.

**Reply by Myndex** [reply]
Hi @lloydk
Thank you for the feedback... can you share the URL or a screenshot?

Right now the maximum of Lc -90 is intended for all text and semantic non-text, and possibly for all elements in dark mode. However, we want to conduct a larger study before saying this as an absolute.

What I imagine at some point, is something like: 

- Lc -90 as a maximum for body text and 1px lines.
- Lc -85 max for text larger than 24px and 2px lines, 
- Lc -80 for very large bold text, and non-text 5px and thicker. 
    - _Lc-75 for large/bold when BG is darker than `#333`_

Again to be clear, these are some initial values that are under study.

**Reply by lloydk** [reply]
Here's a [codepen](https://codepen.io/lloydk/pen/poqbPzX) that shows more more less what the footer looked like. The footer has poor line height and centered text which contributes to the poor readability.

I added an example at 14px with Lc -90 and an example at 20px with a more reasonable line height. I've also added some CSS variables for Lc -85, -80, -75. I find the 20px example at Lc -80 the most pleasant to read.

**Reply by Myndex** [reply]
Hi @lloydk 

As the max contrast is a theory issue, I've started a new discussion at the APCA forum:

https://github.com/Myndex/SAPC-APCA/discussions/106

It would be cool if you posted your last bit with the codepen over there...
